File: github/typescript/src/services/gitData.ts
/**
 * GitHub Git Data Service
 *
 * Low-level operations for Git objects: blobs, trees, commits, references, and tags
 */

export interface RequestOptions {
  timeout?: number;
  headers?: Record<string, string>;
  signal?: AbortSignal;
}

/**
 * Git blob
 */
export interface Blob {
  content: string;
  encoding: 'utf-8' | 'base64';
  url: string;
  sha: string;
  size: number;
  node_id: string;
}

/**
 * Request to create a blob
 */
export interface CreateBlobRequest {
  content: string;
  encoding?: 'utf-8' | 'base64';
}

/**
 * Response from creating a blob
 */
export interface CreateBlobResponse {
  url: string;
  sha: string;
}

/**
 * Git tree entry
 */
export interface TreeEntry {
  path: string;
  mode: '100644' | '100755' | '040000' | '160000' | '120000';
  type: 'blob' | 'tree' | 'commit';
  sha: string;
  size?: number;
  url: string;
}

/**
 * Git tree
 */
export interface Tree {
  sha: string;
  url: string;
  tree: TreeEntry[];
  truncated: boolean;
}

/**
 * Tree entry for creation
 */
export interface CreateTreeEntry {
  path: string;
  mode: '100644' | '100755' | '040000' | '160000' | '120000';
  type: 'blob' | 'tree' | 'commit';
  sha?: string;
  content?: string;
}

/**
 * Request to create a tree
 */
export interface CreateTreeRequest {
  tree: CreateTreeEntry[];
  base_tree?: string;
}

/**
 * Git commit
 */
export interface Commit {
  sha: string;
  node_id: string;
  url: string;
  author: GitUser;
  committer: GitUser;
  message: string;
  tree: {
    sha: string;
    url: string;
  };
  parents: Array<{
    sha: string;
    url: string;
    html_url: string;
  }>;
  verification?: {
    verified: boolean;
    reason: string;
    signature: string | null;
    payload: string | null;
  };
}

/**
 * Git user information
 */
export interface GitUser {
  name: string;
  email: string;
  date: string;
}

/**
 * Request to create a commit
 */
export interface CreateCommitRequest {
  message: string;
  tree: string;
  parents?: string[];
  author?: {
    name: string;
    email: string;
    date?: string;
  };
  committer?: {
    name: string;
    email: string;
    date?: string;
  };
  signature?: string;
}

/**
 * Git reference
 */
export interface GitRef {
  ref: string;
  node_id: string;
  url: string;
  object: {
    type: 'commit' | 'tag' | 'tree' | 'blob';
    sha: string;
    url: string;
  };
}

/**
 * Request to create a reference
 */
export interface CreateRefRequest {
  ref: string;
  sha: string;
}

/**
 * Request to update a reference
 */
export interface UpdateRefRequest {
  sha: string;
  force?: boolean;
}

/**
 * Git tag
 */
export interface GitTag {
  node_id: string;
  tag: string;
  sha: string;
  url: string;
  message: string;
  tagger: GitUser;
  object: {
    type: 'commit' | 'tag' | 'tree' | 'blob';
    sha: string;
    url: string;
  };
  verification?: {
    verified: boolean;
    reason: string;
    signature: string | null;
    payload: string | null;
  };
}

/**
 * Request to create a tag
 */
export interface CreateTagRequest {
  tag: string;
  message: string;
  object: string;
  type: 'commit' | 'tag' | 'tree' | 'blob';
  tagger?: {
    name: string;
    email: string;
    date?: string;
  };
}

/**
 * GitHub Git Data Service
 */
export class GitDataService {
  constructor(
    private readonly baseUrl: string,
    private readonly token: string,
    private readonly owner: string,
    private readonly repo: string
  ) {}

  /**
   * Get a blob by SHA
   */
  async getBlob(sha: string, options?: RequestOptions): Promise<Blob> {
    this.validateSha(sha);
    return this.request<Blob>('GET', `/repos/${this.owner}/${this.repo}/git/blobs/${sha}`, undefined, options);
  }

  /**
   * Create a blob
   */
  async createBlob(request: CreateBlobRequest, options?: RequestOptions): Promise<CreateBlobResponse> {
    if (!request || !request.content) {
      throw new Error('Blob content is required');
    }

    const body = {
      content: request.content,
      encoding: request.encoding || 'utf-8',
    };

    return this.request<CreateBlobResponse>('POST', `/repos/${this.owner}/${this.repo}/git/blobs`, body, options);
  }

  /**
   * Get a tree by SHA
   */
  async getTree(sha: string, recursive?: boolean, options?: RequestOptions): Promise<Tree> {
    this.validateSha(sha);

    const path = recursive
      ? `/repos/${this.owner}/${this.repo}/git/trees/${sha}?recursive=1`
      : `/repos/${this.owner}/${this.repo}/git/trees/${sha}`;

    return this.request<Tree>('GET', path, undefined, options);
  }

  /**
   * Create a tree
   */
  async createTree(request: CreateTreeRequest, options?: RequestOptions): Promise<Tree> {
    this.validateCreateTreeRequest(request);
    return this.request<Tree>('POST', `/repos/${this.owner}/${this.repo}/git/trees`, request, options);
  }

  /**
   * Get a commit by SHA
   */
  async getCommit(sha: string, options?: RequestOptions): Promise<Commit> {
    this.validateSha(sha);
    return this.request<Commit>('GET', `/repos/${this.owner}/${this.repo}/git/commits/${sha}`, undefined, options);
  }

  /**
   * Create a commit
   */
  async createCommit(request: CreateCommitRequest, options?: RequestOptions): Promise<Commit> {
    this.validateCreateCommitRequest(request);
    return this.request<Commit>('POST', `/repos/${this.owner}/${this.repo}/git/commits`, request, options);
  }

  /**
   * List references
   */
  async listRefs(namespace?: string, options?: RequestOptions): Promise<GitRef[]> {
    const path = namespace
      ? `/repos/${this.owner}/${this.repo}/git/refs/${namespace}`
      : `/repos/${this.owner}/${this.repo}/git/refs`;

    return this.request<GitRef[]>('GET', path, undefined, options);
  }

  /**
   * Get a reference
   */
  async getRef(ref: string, options?: RequestOptions): Promise<GitRef> {
    if (!ref) {
      throw new Error('Reference is required');
    }

    // Remove 'refs/' prefix if present
    const cleanRef = ref.startsWith('refs/') ? ref.substring(5) : ref;

    return this.request<GitRef>('GET', `/repos/${this.owner}/${this.repo}/git/refs/${cleanRef}`, undefined, options);
  }

  /**
   * Create a reference
   */
  async createRef(request: CreateRefRequest, options?: RequestOptions): Promise<GitRef> {
    this.validateCreateRefRequest(request);
    return this.request<GitRef>('POST', `/repos/${this.owner}/${this.repo}/git/refs`, request, options);
  }

  /**
   * Update a reference
   */
  async updateRef(ref: string, request: UpdateRefRequest, options?: RequestOptions): Promise<GitRef> {
    if (!ref) {
      throw new Error('Reference is required');
    }

    this.validateSha(request.sha);

    // Remove 'refs/' prefix if present
    const cleanRef = ref.startsWith('refs/') ? ref.substring(5) : ref;

    return this.request<GitRef>(
      'PATCH',
      `/repos/${this.owner}/${this.repo}/git/refs/${cleanRef}`,
      request,
      options
    );
  }

  /**
   * Delete a reference
   */
  async deleteRef(ref: string, options?: RequestOptions): Promise<void> {
    if (!ref) {
      throw new Error('Reference is required');
    }

    // Remove 'refs/' prefix if present
    const cleanRef = ref.startsWith('refs/') ? ref.substring(5) : ref;

    await this.request<void>('DELETE', `/repos/${this.owner}/${this.repo}/git/refs/${cleanRef}`, undefined, options);
  }

  /**
   * Get a tag by SHA
   */
  async getTag(sha: string, options?: RequestOptions): Promise<GitTag> {
    this.validateSha(sha);
    return this.request<GitTag>('GET', `/repos/${this.owner}/${this.repo}/git/tags/${sha}`, undefined, options);
  }

  /**
   * Create a tag object
   */
  async createTag(request: CreateTagRequest, options?: RequestOptions): Promise<GitTag> {
    this.validateCreateTagRequest(request);
    return this.request<GitTag>('POST', `/repos/${this.owner}/${this.repo}/git/tags`, request, options);
  }

  /**
   * Validate SHA format
   */
  private validateSha(sha: string): void {
    if (!sha || typeof sha !== 'string') {
      throw new Error('SHA is required');
    }

    if (!/^[a-f0-9]{40}$/i.test(sha)) {
      throw new Error('Invalid SHA format (must be 40 hex characters)');
    }
  }

  /**
   * Validate create tree request
   */
  private validateCreateTreeRequest(request: CreateTreeRequest): void {
    if (!request) {
      throw new Error('Request is required');
    }

    if (!request.tree || !Array.isArray(request.tree)) {
      throw new Error('Tree entries must be an array');
    }

    if (request.tree.length === 0) {
      throw new Error('Tree entries cannot be empty');
    }

    for (const entry of request.tree) {
      if (!entry.path) {
        throw new Error('Tree entry path is required');
      }

      if (!entry.mode) {
        throw new Error('Tree entry mode is required');
      }

      if (!entry.type) {
        throw new Error('Tree entry type is required');
      }

      if (!entry.sha && !entry.content) {
        throw new Error('Tree entry must have either sha or content');
      }
    }
  }

  /**
   * Validate create commit request
   */
  private validateCreateCommitRequest(request: CreateCommitRequest): void {
    if (!request) {
      throw new Error('Request is required');
    }

    if (!request.message) {
      throw new Error('Commit message is required');
    }

    if (!request.tree) {
      throw new Error('Tree SHA is required');
    }

    this.validateSha(request.tree);

    if (request.parents) {
      if (!Array.isArray(request.parents)) {
        throw new Error('Parents must be an array');
      }

      for (const parent of request.parents) {
        this.validateSha(parent);
      }
    }
  }

  /**
   * Validate create reference request
   */
  private validateCreateRefRequest(request: CreateRefRequest): void {
    if (!request) {
      throw new Error('Request is required');
    }

    if (!request.ref) {
      throw new Error('Reference name is required');
    }

    if (!request.ref.startsWith('refs/')) {
      throw new Error('Reference must start with "refs/"');
    }

    this.validateSha(request.sha);
  }

  /**
   * Validate create tag request
   */
  private validateCreateTagRequest(request: CreateTagRequest): void {
    if (!request) {
      throw new Error('Request is required');
    }

    if (!request.tag) {
      throw new Error('Tag name is required');
    }

    if (!request.message) {
      throw new Error('Tag message is required');
    }

    if (!request.object) {
      throw new Error('Object SHA is required');
    }

    this.validateSha(request.object);

    if (!request.type) {
      throw new Error('Object type is required');
    }

    if (!['commit', 'tag', 'tree', 'blob'].includes(request.type)) {
      throw new Error('Object type must be commit, tag, tree, or blob');
    }
  }

  /**
   * Make an HTTP request
   */
  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    options?: RequestOptions
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const signal = options?.signal || controller.signal;

    const timeout = options?.timeout || 30000;
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(url, {
        method,
        headers: {
          'Accept': 'application/vnd.github+json',
          'Authorization': `Bearer ${this.token}`,
          'X-GitHub-Api-Version': '2022-11-28',
          'Content-Type': 'application/json',
          ...options?.headers,
        },
        body: body ? JSON.stringify(body) : undefined,
        signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const errorBody = await response.json().catch(() => ({}));
        throw new Error(
          errorBody.message || `GitHub API error: ${response.status} ${response.statusText}`
        );
      }

      // DELETE requests may return 204 No Content
      if (response.status === 204) {
        return undefined as T;
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeoutId);
      throw error;
    }
  }
}

File: github/rust/src/services/git_data.rs
//! Git Data operations (blobs, trees, commits, references, tags).

use crate::client::GitHubClient;
use crate::errors::GitHubResult;
use serde::{Deserialize, Serialize};

/// Service for Git Data operations.
pub struct GitDataService<'a> {
    client: &'a GitHubClient,
}

impl<'a> GitDataService<'a> {
    /// Creates a new Git Data service.
    pub fn new(client: &'a GitHubClient) -> Self {
        Self { client }
    }

    // Blob operations

    /// Gets a blob.
    pub async fn get_blob(&self, owner: &str, repo: &str, sha: &str) -> GitHubResult<Blob> {
        self.client
            .get(&format!("/repos/{}/{}/git/blobs/{}", owner, repo, sha))
            .await
    }

    /// Creates a blob.
    pub async fn create_blob(
        &self,
        owner: &str,
        repo: &str,
        request: &CreateBlobRequest,
    ) -> GitHubResult<BlobReference> {
        self.client
            .post(&format!("/repos/{}/{}/git/blobs", owner, repo), request)
            .await
    }

    // Tree operations

    /// Gets a tree.
    pub async fn get_tree(
        &self,
        owner: &str,
        repo: &str,
        sha: &str,
        recursive: bool,
    ) -> GitHubResult<Tree> {
        let path = if recursive {
            format!("/repos/{}/{}/git/trees/{}?recursive=1", owner, repo, sha)
        } else {
            format!("/repos/{}/{}/git/trees/{}", owner, repo, sha)
        };
        self.client.get(&path).await
    }

    /// Creates a tree.
    pub async fn create_tree(
        &self,
        owner: &str,
        repo: &str,
        request: &CreateTreeRequest,
    ) -> GitHubResult<Tree> {
        self.client
            .post(&format!("/repos/{}/{}/git/trees", owner, repo), request)
            .await
    }

    // Commit operations

    /// Gets a commit.
    pub async fn get_commit(&self, owner: &str, repo: &str, sha: &str) -> GitHubResult<GitCommit> {
        self.client
            .get(&format!("/repos/{}/{}/git/commits/{}", owner, repo, sha))
            .await
    }

    /// Creates a commit.
    pub async fn create_commit(
        &self,
        owner: &str,
        repo: &str,
        request: &CreateCommitRequest,
    ) -> GitHubResult<GitCommit> {
        self.client
            .post(&format!("/repos/{}/{}/git/commits", owner, repo), request)
            .await
    }

    // Reference operations

    /// Lists references.
    pub async fn list_refs(
        &self,
        owner: &str,
        repo: &str,
        namespace: Option<&str>,
    ) -> GitHubResult<Vec<GitReference>> {
        let path = if let Some(ns) = namespace {
            format!("/repos/{}/{}/git/refs/{}", owner, repo, ns)
        } else {
            format!("/repos/{}/{}/git/refs", owner, repo)
        };
        self.client.get(&path).await
    }

    /// Gets a reference.
    pub async fn get_ref(
        &self,
        owner: &str,
        repo: &str,
        ref_name: &str,
    ) -> GitHubResult<GitReference> {
        // Remove leading "refs/" if present
        let ref_path = ref_name.strip_prefix("refs/").unwrap_or(ref_name);
        self.client
            .get(&format!("/repos/{}/{}/git/refs/{}", owner, repo, ref_path))
            .await
    }

    /// Creates a reference.
    pub async fn create_ref(
        &self,
        owner: &str,
        repo: &str,
        ref_name: &str,
        sha: &str,
    ) -> GitHubResult<GitReference> {
        let request = CreateRefRequest {
            ref_name: if ref_name.starts_with("refs/") {
                ref_name.to_string()
            } else {
                format!("refs/{}", ref_name)
            },
            sha: sha.to_string(),
        };
        self.client
            .post(&format!("/repos/{}/{}/git/refs", owner, repo), &request)
            .await
    }

    /// Updates a reference.
    pub async fn update_ref(
        &self,
        owner: &str,
        repo: &str,
        ref_name: &str,
        sha: &str,
        force: bool,
    ) -> GitHubResult<GitReference> {
        // Remove leading "refs/" if present
        let ref_path = ref_name.strip_prefix("refs/").unwrap_or(ref_name);
        let request = UpdateRefRequest {
            sha: sha.to_string(),
            force,
        };
        self.client
            .patch(
                &format!("/repos/{}/{}/git/refs/{}", owner, repo, ref_path),
                &request,
            )
            .await
    }

    /// Deletes a reference.
    pub async fn delete_ref(&self, owner: &str, repo: &str, ref_name: &str) -> GitHubResult<()> {
        // Remove leading "refs/" if present
        let ref_path = ref_name.strip_prefix("refs/").unwrap_or(ref_name);
        self.client
            .delete(&format!("/repos/{}/{}/git/refs/{}", owner, repo, ref_path))
            .await
    }

    // Tag operations

    /// Gets a tag.
    pub async fn get_tag(&self, owner: &str, repo: &str, sha: &str) -> GitHubResult<GitTag> {
        self.client
            .get(&format!("/repos/{}/{}/git/tags/{}", owner, repo, sha))
            .await
    }

    /// Creates a tag.
    pub async fn create_tag(
        &self,
        owner: &str,
        repo: &str,
        request: &CreateTagRequest,
    ) -> GitHubResult<GitTag> {
        self.client
            .post(&format!("/repos/{}/{}/git/tags", owner, repo), request)
            .await
    }
}

// Blob types

/// A Git blob.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Blob {
    /// The SHA of the blob.
    pub sha: String,
    /// The node ID.
    pub node_id: String,
    /// The size of the blob in bytes.
    pub size: u64,
    /// The URL of the blob.
    pub url: String,
    /// The content of the blob (base64 encoded or plain text).
    pub content: Option<String>,
    /// The encoding of the content.
    pub encoding: String,
}

/// A reference to a created blob.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlobReference {
    /// The SHA of the created blob.
    pub sha: String,
    /// The URL of the blob.
    pub url: String,
}

/// Request to create a blob.
#[derive(Debug, Clone, Serialize)]
pub struct CreateBlobRequest {
    /// The content of the blob.
    pub content: String,
    /// The encoding of the content (utf-8 or base64).
    pub encoding: BlobEncoding,
}

/// Blob encoding.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum BlobEncoding {
    /// UTF-8 encoding.
    #[serde(rename = "utf-8")]
    Utf8,
    /// Base64 encoding.
    Base64,
}

// Tree types

/// A Git tree.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tree {
    /// The SHA of the tree.
    pub sha: String,
    /// The URL of the tree.
    pub url: String,
    /// The tree entries.
    pub tree: Vec<TreeEntry>,
    /// Whether the tree was truncated.
    pub truncated: bool,
}

/// A tree entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TreeEntry {
    /// The path of the entry.
    pub path: String,
    /// The file mode.
    pub mode: String,
    /// The type of entry (blob, tree, commit).
    #[serde(rename = "type")]
    pub entry_type: String,
    /// The SHA of the entry.
    pub sha: Option<String>,
    /// The size in bytes (for blobs).
    pub size: Option<u64>,
    /// The URL of the entry.
    pub url: Option<String>,
}

/// Request to create a tree.
#[derive(Debug, Clone, Serialize)]
pub struct CreateTreeRequest {
    /// The tree entries to create.
    pub tree: Vec<CreateTreeEntry>,
    /// The SHA of the base tree (optional).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_tree: Option<String>,
}

/// An entry to create in a tree.
#[derive(Debug, Clone, Serialize)]
pub struct CreateTreeEntry {
    /// The file path.
    pub path: String,
    /// The file mode.
    pub mode: TreeMode,
    /// The type of entry.
    #[serde(rename = "type")]
    pub entry_type: TreeEntryType,
    /// The SHA of the blob/tree/commit (mutually exclusive with content).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sha: Option<String>,
    /// The content (mutually exclusive with sha).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
}

/// Tree entry mode.
#[derive(Debug, Clone, Serialize)]
pub enum TreeMode {
    /// Regular file (100644).
    #[serde(rename = "100644")]
    File,
    /// Executable file (100755).
    #[serde(rename = "100755")]
    Executable,
    /// Subdirectory (040000).
    #[serde(rename = "040000")]
    Subdirectory,
    /// Submodule (160000).
    #[serde(rename = "160000")]
    Submodule,
    /// Symbolic link (120000).
    #[serde(rename = "120000")]
    Symlink,
}

/// Tree entry type.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum TreeEntryType {
    /// A blob (file).
    Blob,
    /// A tree (directory).
    Tree,
    /// A commit (submodule).
    Commit,
}

// Commit types

/// A Git commit.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitCommit {
    /// The SHA of the commit.
    pub sha: String,
    /// The node ID.
    pub node_id: String,
    /// The URL of the commit.
    pub url: String,
    /// The author of the commit.
    pub author: GitUser,
    /// The committer of the commit.
    pub committer: GitUser,
    /// The commit message.
    pub message: String,
    /// The tree information.
    pub tree: TreeReference,
    /// The parent commits.
    pub parents: Vec<ParentCommit>,
    /// Verification information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verification: Option<Verification>,
}

/// Git user information.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitUser {
    /// The name of the user.
    pub name: String,
    /// The email of the user.
    pub email: String,
    /// The date (ISO 8601 format).
    pub date: String,
}

/// Tree reference in a commit.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TreeReference {
    /// The SHA of the tree.
    pub sha: String,
    /// The URL of the tree.
    pub url: String,
}

/// Parent commit reference.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParentCommit {
    /// The SHA of the parent commit.
    pub sha: String,
    /// The URL of the parent commit.
    pub url: String,
    /// The HTML URL of the parent commit.
    pub html_url: String,
}

/// Verification information for signed commits.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Verification {
    /// Whether the commit was verified.
    pub verified: bool,
    /// The reason for the verification status.
    pub reason: String,
    /// The signature.
    pub signature: Option<String>,
    /// The payload that was signed.
    pub payload: Option<String>,
}

/// Request to create a commit.
#[derive(Debug, Clone, Serialize)]
pub struct CreateCommitRequest {
    /// The commit message.
    pub message: String,
    /// The SHA of the tree.
    pub tree: String,
    /// The SHAs of the parent commits.
    pub parents: Vec<String>,
    /// The author information (optional).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<CommitAuthor>,
    /// The committer information (optional).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub committer: Option<CommitAuthor>,
    /// The PGP signature (optional).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<String>,
}

/// Commit author/committer information.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommitAuthor {
    /// The name of the author/committer.
    pub name: String,
    /// The email of the author/committer.
    pub email: String,
    /// The timestamp (ISO 8601 format, optional).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub date: Option<String>,
}

// Reference types

/// A Git reference.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitReference {
    /// The reference name (e.g., "refs/heads/main").
    #[serde(rename = "ref")]
    pub ref_name: String,
    /// The node ID.
    pub node_id: String,
    /// The URL of the reference.
    pub url: String,
    /// The object the reference points to.
    pub object: GitObject,
}

/// A Git object.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitObject {
    /// The SHA of the object.
    pub sha: String,
    /// The type of object (commit, tree, blob, tag).
    #[serde(rename = "type")]
    pub object_type: String,
    /// The URL of the object.
    pub url: String,
}

/// Request to create a reference.
#[derive(Debug, Clone, Serialize)]
pub struct CreateRefRequest {
    /// The reference name (must start with "refs/").
    #[serde(rename = "ref")]
    pub ref_name: String,
    /// The SHA to point the reference to.
    pub sha: String,
}

/// Request to update a reference.
#[derive(Debug, Clone, Serialize)]
pub struct UpdateRefRequest {
    /// The new SHA for the reference.
    pub sha: String,
    /// Whether to force the update.
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub force: bool,
}

// Tag types

/// A Git tag.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitTag {
    /// The SHA of the tag.
    pub sha: String,
    /// The node ID.
    pub node_id: String,
    /// The URL of the tag.
    pub url: String,
    /// The tag name.
    pub tag: String,
    /// The tag message.
    pub message: String,
    /// The tagger information.
    pub tagger: GitUser,
    /// The object the tag points to.
    pub object: GitObject,
    /// Verification information.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verification: Option<Verification>,
}

/// Request to create a tag.
#[derive(Debug, Clone, Serialize)]
pub struct CreateTagRequest {
    /// The tag name.
    pub tag: String,
    /// The tag message.
    pub message: String,
    /// The SHA of the object to tag.
    pub object: String,
    /// The type of object (commit, tree, blob).
    #[serde(rename = "type")]
    pub object_type: TagObjectType,
    /// The tagger information (optional).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tagger: Option<CommitAuthor>,
}

/// Tag object type.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum TagObjectType {
    /// A commit.
    Commit,
    /// A tree.
    Tree,
    /// A blob.
    Blob,
}
